storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -0,0 +1,873 @@
+            }
+        };
+        LOG_FLUSH_TIMER = logFlushStatsMetricsGroup.newTimer("LogFlushRateAndTimeMs", TimeUnit.MILLISECONDS, TimeUnit.SECONDS);
+    }
+
+    private final FileRecords log;
+    private final LazyIndex<OffsetIndex> lazyOffsetIndex;
+    private final LazyIndex<TimeIndex> lazyTimeIndex;
+    private final TransactionIndex txnIndex;
+    private final long baseOffset;
+    private final int indexIntervalBytes;
+    private final long rollJitterMs;
+    private final Time time;
+
+    // The timestamp we used for time based log rolling and for ensuring max compaction delay
+    // volatile for LogCleaner to see the update
+    private volatile OptionalLong rollingBasedTimestamp = OptionalLong.empty();
+
+    /* The maximum timestamp and offset we see so far */
+    private volatile TimestampOffset maxTimestampAndOffsetSoFar = TimestampOffset.UNKNOWN;
+
+    private long created;
+
+    /* the number of bytes since we last added an entry in the offset index */
+    private int bytesSinceLastIndexEntry = 0;
+
+    /**
+     * Create a LogSegment with the provided parameters.
+     *
+     * @param log The file records containing log entries
+     * @param lazyOffsetIndex The offset index
+     * @param lazyTimeIndex The timestamp index
+     * @param txnIndex The transaction index
+     * @param baseOffset A lower bound on the offsets in this segment
+     * @param indexIntervalBytes The approximate number of bytes between entries in the index
+     * @param rollJitterMs The maximum random jitter subtracted from the scheduled segment roll time
+     * @param time The time instance
+     */
+    public LogSegment(FileRecords log,
+                      LazyIndex<OffsetIndex> lazyOffsetIndex,
+                      LazyIndex<TimeIndex> lazyTimeIndex,
+                      TransactionIndex txnIndex,
+                      long baseOffset,
+                      int indexIntervalBytes,
+                      long rollJitterMs,
+                      Time time) {
+        this.log = log;
+        this.lazyOffsetIndex = lazyOffsetIndex;
+        this.lazyTimeIndex = lazyTimeIndex;
+        this.txnIndex = txnIndex;
+        this.baseOffset = baseOffset;
+        this.indexIntervalBytes = indexIntervalBytes;
+        this.rollJitterMs = rollJitterMs;
+        this.time = time;
+        this.created = time.milliseconds();
+    }
+
+    // Visible for testing
+    public LogSegment(LogSegment segment) {
+        this(segment.log, segment.lazyOffsetIndex, segment.lazyTimeIndex, segment.txnIndex, segment.baseOffset,
+                segment.indexIntervalBytes, segment.rollJitterMs, segment.time);
+    }
+
+    public OffsetIndex offsetIndex() throws IOException {
+        return lazyOffsetIndex.get();
+    }
+
+    public File offsetIndexFile() {
+        return lazyOffsetIndex.file();
+    }
+
+    public TimeIndex timeIndex() throws IOException {
+        return lazyTimeIndex.get();
+    }
+
+    public File timeIndexFile() {
```

Comment:
It's a public because it needs to be a public method. What you suggested doesn't work because it forces materialization and it would be a serious regression.

Reply:
We are leaking the fact here that the index is implemented using a lazy index implementation outside the segment.

Does the caller know that File returned by `timeIndexFile()` may not be consistent with the in-memory state of index? We are adding the responsibility on the caller to ensure that it calls flush() if it requires consistency. This means that we are leaking the internal implementation of index (being lazy) to the caller. This is concerning because it can cause bugs where authors using this method in other parts of the code may not understand that it could be eventually inconsistent.

If we really want to provide an index reference which doesn't require materialization, why not share the LazyIndex with the caller. Then the caller explicitly knows that this index is lazily evaluated?

Also, where are we using this function outside this class? Should this be private?

Reply:
We are not leaking anything that is not already exposed. It would be different if that was not accessible via the overall interface already. The only discussion here is not whether this is exposed (it already is), but _how_ to expose it.

Reply:
> We are not leaking anything that is not already exposed. It would be different if that was not accessible via the overall interface already. The only discussion here is not whether this is exposed (it already is), but how to expose it.

Fair point. Let tackle the how question in this PR and take up plugging the underlying file implementation leak separately later.

On the how part, can we keep the index or LaxyIndex as the components who choose to expose internal implementation outside (instead of LogSegment)? What that means is that we will probably have a function here in LogSegment which returns LazyIndex.

Thoughts?

Reply:
`LazyIndex` exposes a larger surface area though. And it's confusing to expose both `LazyIndex` and the materialized index. The way it is now is actually simpler. Local indexes are file based and that's core to how they work. `LogSegment` gives you the file if that's all you need or the materialized index. And `LazyIndex` is used internally to simplify the implementation.

Reply:
ok, could we add a javadoc on this method that the physical file pointed to by this file object may not be consistent with in-memory copy of index? It doesn't totally address my concern of a potential bug when someone uses this File and assumes that it represents a consistent view of the index but I will am happy with a javadoc now. The whole index implemented by a file thing needs to be hidden away but that belongs in a separate JIRA.

Reply:
The indexes are memory mapped, so in theory it should be consistent for reads that go via the page cache.

Reply:
One last clarification: I am not opposed to improving the overall modularity of the classes in the storage layer. As you said, in a separate PR/discussion.